### DISCUSSION TITLE
Possible to choose longhorn storage type?

### DISCUSSION THREAD

**Discussion by andinger** [mysticaltech/terraform-hcloud-kube-hetzner#373] [Q&A] [answered]
Hi, firstly thx for this amazing project!
I have a question regarding longhorn volumes. Is it possible to decide at the moment of the concrete volume declaration, if longhorn creates the volume on the node or block storage?

Background: i want to use longhorn volumes for database persistence (should be internal ssd on the node) and a block storage volume for a big amount of files.

Thx and regards, Andi

**Comment by ifeulner** [comment]
Yes, this is possible. You need to define the storage locations and dedicated StorageClasses that use then this storage location(e.g. "nvme", "ssd"). This is done by a specific label and annotation on the nodes.

**Reply by mysticaltech** [reply]
Also, @andinger, please look at this explanation, indeed, using both kinds is possible... https://github.com/kube-hetzner/terraform-hcloud-kube-hetzner/issues/369#issuecomment-1289597178

**Comment by andinger** [comment]
wow thx for you fast replies ... just to check, if i got your points correctly ...

* i set the `longhorn_volume_size` on the database nodes to create node volumes and label these nodes e.g. as db-nodes
* for the file (block storage) volumes, i need to set `longhorn_volume_size=0` on the other volumes and label e.g. as storage-nodes
* i create two StorageClasses e.g. as shown in https://longhorn.io/kb/tip-only-use-storage-on-a-set-of-nodes/ 
* when i create the concrete volumes, i select the correct StorageClass

**Reply by mysticaltech** [reply]
Hmm, @andinger, I believe there is a confusion. I will recap:

We are talking about agent node pool definitions here, see the kube.tf.example.

All agent node pool definitions need to have the same attributes, but for each node pool those can have different values.

- So for node pools that you do not want to be involved in any kind of storage activity with longhorn, you set the taint as `server-usage=storage:NoSchedule`, with an attribute of `longhorn_volume_size=0`.

- For nodes you want to use for DBs, you have to use Fast longhorn storage, that is only achievable via node storage, so just set `longhorn_volume_size=0`, and you will have the full storage of the node minus 20% reserved for the OS.

- For nodes where you need low IOPS speeds, so "slow storage" you can use mounted volume storage that is cheaper, for those node pools, you set `longhorn_volume_size=xxx`. This will ensure storage is done on a mounted volume; this is not ok for DBs, but probably ok for S3-type storage IMHO.

Also, in choosing the node types that you will use with Longhorn for storage, choose at least node types with 4 CPUs and 4 GB of ram minimum, as per https://longhorn.io/docs/1.3.2/best-practices/.

@ifeulner @WolfspiritM Did I miss something here?

**Reply by mysticaltech** [reply]
@andinger Also good to know, it seems that the latest version of Longhorn is not yet compatible with v1.25 kube, so make sure to set in your kube.tf a value for `initial_k3s_channel`  = "v1.24".

**Reply by mkreuzmayr** [reply]
If I enable longhorn for all nodes, is it possible to decide on a per deployment basis (StorageClass) if it should use node or volume storage?

**Reply by mysticaltech** [reply]
Everyone please note that `server-usage=storage:NoSchedule` has been removed from the docs, because it turns out that it just blocks scheduling for all pods, so not ideal.

**Reply by mysticaltech** [reply]
@mkreuzmayr Absolutely... For volumes, set `longhorn_volume_size` > 10, and for node storage, set it to `longhorn_volume_size=0`. You do not even need anything else.

**Reply by mkreuzmayr** [reply]
@mysticaltech Yes but as far as I understood, that needs to be defined on the nodepool or am I wrong?

My question is if I have 3 nodes, all with longhorn volumes, is it possible to specify on a `StorageClass` which type of storage I want to use for my persistent volume. My plan is to have some containers using volumes and some the node storage directly.

**Reply by ifeulner** [reply]
Sure, this is a standard K8s topic, the PersistentVolumeClaim (PVC) definition refers to the StorageClass you want to use. This could be either a longhorn-based SC or and hcloud-based SC (or any other, depending on your cluster setup)

**Reply by mkreuzmayr** [reply]
Ok, I think I mixed things a bit, so longhorn only uses node storage and for volume storage i have to use hetzner CSI, am I right?

**Reply by ifeulner** [reply]
No, you can use either longhorn or hetzner, as they both implement the Container Storage Interface (CSI) which provides storage for containers.
But the Hetzner CSI only uses hcloud volumes, while longhorn uses the (local) hcloud-server disk (which is actually an NVMe I think).

**Reply by mkreuzmayr** [reply]
Yes, that's exactly what I meant.... Just in other words. 
I'm fairly new to k8s and have only used it to host some HA containers, but haven't really bothered with storage.
Thanks for your quick explanations & help.

**Reply by mysticaltech** [reply]
The only thing to note @mkreuzmayr, and that is very important to understand. A node (meaning nodepools) cannot have volume storage or node storage at the same time, `longhorn_volume_size` as explained above is the discriminator. 

Now, how all this works with PVCs, personally I have no idea. So maybe the nodepools using volumes (meaning `longhorn_volume_size` must be tagged with an appropriate label of "volume-storage") and the same thing for node storage. 

Then maybe via longhorn config you can associate these labels with their respective StorageClass. I hope that is clearer by now, and if you find the full flow, please do share it here for everyone to learn.

**Reply by mysticaltech** [reply]
Also important to remember, volumes are slow and node-storage is fast but tends to be more expensive of course. So for DBs for instance, node storage seems to be the appropriate medium.

**Reply by gefloh** [reply]
> The only thing to note @mkreuzmayr, and that is very important to understand. A node (meaning nodepools) cannot have volume storage or node storage at the same time, `longhorn_volume_size` as explained above is the discriminator.

But is this a general limitation of Longhorn @mysticaltech ? Or is it just not possible to provision currently via Terraform? Because it would mean if I use volume storage on a node pool, I would lose all the local node storage for these nodes (or at least would not be able to use it for storage) or am I misunderstanding that?

**Reply by ifeulner** [reply]
@gefloh @mysticaltech Actually, both is possible, you can use the node's local storage  AND also additionally mounted hetzner volumes  for longhorn storage. See an example here: https://gist.github.com/ifeulner/d311b2868f6c00e649f33a72166c2e5b#file-01_longhorn_bestpractices-md

**Reply by mysticaltech** [reply]
@gefloh It's how we designed the system, but @ifeulner seems to have found a way, so I would recommend following his excellant guide!

**Reply by gefloh** [reply]
Thanks a lot, I think I got the idea what to do. Just two more questions @ifeulner : 

- At the very beginning under "Initial configuration" there is this yaml. Where do I find this/how can I set this as a label?
- Is this configuration only for newly created nodes or can I apply it also to existing ones (assuming that there are no volumes defined)?

**Reply by mysticaltech** [reply]
@gefloh Just create a cluster with enable_longhorn and the volume size you need. Then when the cluster is up, you can create, apply new yamls related to longhorn. And if you need to fetch one from the cluster you can too. Really depends what you want to do.

**Reply by gefloh** [reply]
Ok I misunderstood this sentence under "Initial configuration". The yaml is probably not for labeling nodes, it's most probably just the default settings. What confuses me still is that according to the Longhorn docs, the value is named `create-default-disk-labeled-nodes`. 
@mysticaltech What I want to do is use the local node storage of all worker nodes but still have one or two additional Hetzner volumes attached to one or two nodes for stuff where performance doesn't matter. Will continue experimenting.

**Reply by gefloh** [reply]
Ah ok, sorry for the confusion, now I found it, these are the values for the Longhorn Helm chart to be set in the kube.tf.

**Reply by gefloh** [reply]
Works like a charm, thanks a lot @ifeulner and @mysticaltech !

**Reply by garry-t** [reply]
@gefloh could you please explain how you achieved this? I couldn't figure out. I also want to allow use by default local storage of node and only for some node longhorn..

**Comment by andinger** [comment]
Hey guys, thanks for your support – i have been on the road quite a bit in the last two weeks so i didn't have time to check your tips. regardless ... is there a way i can sponsor this project? would love to give something back to the dev community ...

**Reply by mysticaltech** [reply]
@andinger Thanks for this kind proposal, I will add a sponsors option ASAP and let you know! 🙏 💙

**Comment by andinger** [comment, marked as answer]
Hey guys, i finally got it working.

* create local storage node pool with `longhorn_volume_size=0`
* create node pool with a specific longhorn volume size to be created as hetzner block volumes
* set annotations to the concrete nodes of the pools, e.g. `node.longhorn.io/default-node-tags: ["local"]` (btw. would be great if i could set the default node annotations directly in the node pool definition)
* create two longhorn based storage classes with the corresponding node selector in the parameters (https://longhorn.io/kb/tip-only-use-storage-on-a-set-of-nodes/#create-a-storageclass-that-select-a-specific-set-of-nodes)

Thx for your support!

**Reply by ifeulner** [reply]
Great - there is also more configuration possibilities via annotations: This might help here (WIP): https://gist.github.com/ifeulner/d311b2868f6c00e649f33a72166c2e5b
